File: src/python/systems/validate_phase10_freezes.py
"""Independently validate the final Sol-accepted Phase 10 freeze manifests."""
from __future__ import annotations

import hashlib
import json
import re
import subprocess
from pathlib import Path

from freeze_hash import manifest_matches

ROOT = Path(__file__).resolve().parents[3]
REPORTS = ROOT / "reports"
SOURCE_COMMIT = "e1f233cc89e3694d2f08dfd81fde6f9194f57a65"
FINAL = {
    "10A": {
        "path": REPORTS / "phase10a_governance_jurisdiction_freeze_manifest.json",
        "baseline": "Governance & Jurisdiction Baseline, 2026",
        "counts": {"actors": 40, "authorities": 100, "relationships": 100, "sources": 48, "uncertainties": 16},
        "artifacts": {
            "data/processed/analysis/governance_actors.csv",
            "data/processed/analysis/governance_authorities.csv",
            "data/processed/networks/governance_relationships.csv",
            "data/processed/analysis/governance_sources.csv",
            "data/processed/analysis/governance_uncertainty_register.csv",
            "outputs/maps/systems/32_governance_jurisdiction_2026.png",
            "outputs/maps/systems/32_governance_jurisdiction_2026.svg",
            "reports/governance_baseline_sources.md",
            "reports/governance_baseline_assumptions.md",
            "reports/governance_baseline_findings.md",
            "reports/governance_baseline_qa.md",
            "reports/governance_baseline_artifact_check.json",
            "reports/governance_independent_review.md",
            "reports/governance_post_correction_independent_review.md",
            "reports/governance_additional_post_correction_independent_review.md",
        },
    },
    "10B": {
        "path": REPORTS / "phase10b_governance_dependencies_coordination_freeze_manifest.json",
        "baseline": "Cross-System Authority, Dependencies & Coordination, 2026",
        "counts": {"dependency_register": 25, "dependency_edges": 25, "coordination_mechanisms": 14, "matrix_rows": 10, "sources_reused": 48},
        "artifacts": {
            "data/processed/analysis/governance_dependency_register.csv",
            "data/processed/analysis/governance_coordination_mechanisms.csv",
            "data/processed/analysis/governance_dependency_matrix.csv",
            "data/processed/networks/governance_dependency_edges.csv",
            "data/processed/analysis/governance_sources.csv",
            "outputs/maps/systems/33_cross_system_governance_dependencies_2026.png",
            "outputs/maps/systems/33_cross_system_governance_dependencies_2026.svg",
            "reports/governance_dependency_sources.md",
            "reports/governance_dependency_assumptions.md",
            "reports/governance_dependency_findings.md",
            "reports/governance_dependency_qa.md",
            "reports/governance_dependency_artifact_check.json",
            "reports/governance_independent_review.md",
            "reports/governance_post_correction_independent_review.md",
            "reports/governance_additional_post_correction_independent_review.md",
        },
    },
}


def digest(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def artifact_matches(path: Path, metadata: dict[str, object]) -> bool:
    data = path.read_bytes()
    expected = str(metadata["sha256"])
    if path.suffix.lower() not in {".csv", ".json", ".md", ".txt", ".yml", ".yaml", ".svg"}:
        return len(data) == int(metadata["bytes"]) and hashlib.sha256(data).hexdigest() == expected
    canonical = data.replace(b"\r\n", b"\n").replace(b"\r", b"\n")
    crlf = canonical.replace(b"\n", b"\r\n")
    candidates = {hashlib.sha256(data).hexdigest(), hashlib.sha256(canonical).hexdigest(), hashlib.sha256(crlf).hexdigest()}
    return expected in candidates and int(metadata["bytes"]) in {len(data), len(canonical), len(crlf)}


def verify_final_manifests() -> tuple[int, int, set[str]]:
    seen: set[str] = set()
    entries = 0
    for phase, spec in FINAL.items():
        manifest = json.loads(spec["path"].read_text(encoding="utf-8"))
        assert manifest["accepted_phase"] == phase
        assert manifest["baseline"] == spec["baseline"]
        assert manifest["status"] == "ACCEPTED / FROZEN"
        assert manifest["accepted_by"] == "Sol explicit acceptance decision supplied for this run"
        assert manifest["accepted_date"] == "2026-09-04"
        assert manifest["source_commit"] == SOURCE_COMMIT
        assert manifest["counts"] == spec["counts"]
        artifacts = manifest["artifacts"]
        assert set(artifacts) == spec["artifacts"], phase
        for rel, metadata in artifacts.items():
            path = ROOT / rel
            assert path.exists(), rel
            assert set(metadata) == {"sha256", "bytes", "role"}, rel
            assert artifact_matches(path, metadata), rel
            seen.add(rel)
            entries += 1
    b = json.loads(FINAL["10B"]["path"].read_text(encoding="utf-8"))
    a_path = ROOT / b["phase10a_freeze_manifest"]["path"]
    assert b["phase10a_freeze_manifest"]["sha256"] == digest(a_path)
    assert b["phase10a_freeze_manifest"]["bytes"] == a_path.stat().st_size
    return entries, len(seen), seen


def verify_prior_freezes(final_artifacts: set[str]) -> tuple[int, int]:
    prior_paths: set[str] = set()
    total = 0
    newline_only = 0
    final_names = {spec["path"].name for spec in FINAL.values()}
    for path in sorted(REPORTS.glob("phase*_freeze_manifest.json")):
        if path.name in final_names:
            continue
        manifest = json.loads(path.read_text(encoding="utf-8"))
        artifacts = manifest.get("artifacts", manifest.get("files", {}))
        for rel, metadata in artifacts.items():
            artifact = ROOT / rel
            expected = metadata["sha256"] if isinstance(metadata, dict) else metadata
            assert artifact.exists(), rel
            if isinstance(metadata, dict) and "bytes" in metadata:
                assert artifact.stat().st_size == metadata["bytes"], rel
            if digest(artifact) != expected:
                assert manifest_matches(ROOT, rel, expected), rel
                newline_only += 1
            assert rel not in prior_paths, f"prior artifact assigned twice: {rel}"
            assert rel not in final_artifacts, f"Phase 1-9 artifact overlaps Phase 10: {rel}"
            prior_paths.add(rel)
            total += 1
    assert total == 196, total
    changed = subprocess.check_output(["git", "-C", str(ROOT), "diff", "--name-only", SOURCE_COMMIT], text=True).splitlines()
    assert not set(changed).intersection(prior_paths)
    return total, newline_only


def verify_records_and_status() -> None:
    post = (REPORTS / "governance_post_correction_independent_review.md").read_text(encoding="utf-8")
    additional = (REPORTS / "governance_additional_post_correction_independent_review.md").read_text(encoding="utf-8")
    assert "deleg_2e8d515e" in post and '"passed": true' in post
    assert "deleg_eedc4116" in additional and '"passed": true' in additional
    for rel in (
        "PROJECT_STATUS.md",
        "docs/canon_status.md",
        "docs/phase_briefs/phase10a_governance_jurisdiction_baseline.md",
        "docs/phase_briefs/phase10b_governance_dependencies_coordination.md",
        "reports/current_phase_handoff.md",
        "README.md",
        "CHANGELOG.md",
        "reports/README.md",
    ):
        text = (ROOT / rel).read_text(encoding="utf-8").lower()
        assert "accepted / frozen" in text, rel
        assert "phase 10c" in text and "not implemented" in text, rel
    canon = (ROOT / "docs/canon_status.md").read_text(encoding="utf-8")
    handoff = (ROOT / "reports/current_phase_handoff.md").read_text(encoding="utf-8")
    for text in (canon, handoff):
        lower = text.lower()
        assert "great black swamp" in lower and "hold" in lower and "noncanonical" in lower
        assert "intake-coordinate discrepancy" in lower and "unresolved" in lower
    changed = subprocess.check_output(["git", "-C", str(ROOT), "diff", "--name-only", SOURCE_COMMIT], text=True).splitlines()
    assert all("phase10c" not in rel.lower() or rel.lower() == "docs/phase_briefs/phase10c_governance_futures.md" for rel in changed)


def main() -> None:
    entries, unique, final_artifacts = verify_final_manifests()
    prior, newline_only = verify_prior_freezes(final_artifacts)
    verify_records_and_status()
    result = {
        "status": "passed",
        "phase10_manifests": 2,
        "phase10_manifest_entries": entries,
        "phase10_unique_artifacts": unique,
        "prior_phase1_9_protected_artifacts": prior,
        "prior_newline_only_matches": newline_only,
        "manifest_hashes_valid": True,
        "status_consistency_valid": True,
        "review_records_preserved": ["deleg_2e8d515e", "deleg_eedc4116"],
        "phase10c_unimplemented": True,
        "holds_preserved": True,
    }
    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: src/R/systems/validate_phase10_freezes.R
args <- commandArgs(trailingOnly=TRUE)
root <- if (length(args)) args[[1]] else normalizePath(file.path(dirname(sys.frame(1)$ofile), "../../.."), mustWork=FALSE)
reports <- file.path(root, "reports")
source_commit <- "e1f233cc89e3694d2f08dfd81fde6f9194f57a65"
final_names <- c(
  "phase10a_governance_jurisdiction_freeze_manifest.json",
  "phase10b_governance_dependencies_coordination_freeze_manifest.json"
)
expected_phase <- c("10A", "10B")
expected_counts <- list(
  c(actors=40L, authorities=100L, relationships=100L, sources=48L, uncertainties=16L),
  c(dependency_register=25L, dependency_edges=25L, coordination_mechanisms=14L, matrix_rows=10L, sources_reused=48L)
)
expected_artifacts <- list(
  c(
    "data/processed/analysis/governance_actors.csv",
    "data/processed/analysis/governance_authorities.csv",
    "data/processed/networks/governance_relationships.csv",
    "data/processed/analysis/governance_sources.csv",
    "data/processed/analysis/governance_uncertainty_register.csv",
    "outputs/maps/systems/32_governance_jurisdiction_2026.png",
    "outputs/maps/systems/32_governance_jurisdiction_2026.svg",
    "reports/governance_baseline_sources.md",
    "reports/governance_baseline_assumptions.md",
    "reports/governance_baseline_findings.md",
    "reports/governance_baseline_qa.md",
    "reports/governance_baseline_artifact_check.json",
    "reports/governance_independent_review.md",
    "reports/governance_post_correction_independent_review.md",
    "reports/governance_additional_post_correction_independent_review.md"
  ),
  c(
    "data/processed/analysis/governance_dependency_register.csv",
    "data/processed/analysis/governance_coordination_mechanisms.csv",
    "data/processed/analysis/governance_dependency_matrix.csv",
    "data/processed/networks/governance_dependency_edges.csv",
    "data/processed/analysis/governance_sources.csv",
    "outputs/maps/systems/33_cross_system_governance_dependencies_2026.png",
    "outputs/maps/systems/33_cross_system_governance_dependencies_2026.svg",
    "reports/governance_dependency_sources.md",
    "reports/governance_dependency_assumptions.md",
    "reports/governance_dependency_findings.md",
    "reports/governance_dependency_qa.md",
    "reports/governance_dependency_artifact_check.json",
    "reports/governance_independent_review.md",
    "reports/governance_post_correction_independent_review.md",
    "reports/governance_additional_post_correction_independent_review.md"
  )
)

sha256_file <- function(path) {
  cmd <- Sys.which("sha256sum")
  if (nchar(cmd) > 0) {
    out <- system2(cmd, path, stdout=TRUE, stderr=TRUE)
    stopifnot(length(out) >= 1)
    token <- strsplit(trimws(out[[1]]), "[[:space:]]+")[[1]][1]
    return(tolower(gsub("[^0-9a-f]", "", token)))
  }
  certutil <- Sys.which("certutil")
  stopifnot(nchar(certutil) > 0)
  out <- system2(certutil, c("-hashfile", path, "SHA256"), stdout=TRUE, stderr=TRUE)
  normalized <- tolower(gsub("[[:space:]]+", "", out))
  hits <- normalized[grepl("^[0-9a-f]{64}$", normalized)]
  stopifnot(length(hits) >= 1)
  hits[[1]]
}

manifest_artifacts <- function(path) {
  lines <- readLines(path, warn=FALSE)
  current <- NA_character_; current_bytes <- NA_real_; current_hash <- NA_character_; rel <- character(); bytes <- numeric(); hashes <- character()
  append_current <- function() {
    if (!is.na(current) && !is.na(current_hash)) {
      rel <<- c(rel, current); bytes <<- c(bytes, current_bytes); hashes <<- c(hashes, current_hash)
      current <<- NA_character_; current_bytes <<- NA_real_; current_hash <<- NA_character_
    }
  }
  for (line in lines) {
    direct <- regexec('^[[:space:]]*"([^"]+)"[[:space:]]*:[[:space:]]*"([0-9a-f]{64})"', line, perl=TRUE)
    direct_hit <- regmatches(line, direct)[[1]]
    if (length(direct_hit) == 3 && grepl("/", direct_hit[[2]], fixed=TRUE)) {
      rel <- c(rel, direct_hit[[2]]); bytes <- c(bytes, NA_real_); hashes <- c(hashes, direct_hit[[3]])
      next
    }
    m <- regexec('^[[:space:]]*"([^"]+)"[[:space:]]*:[[:space:]]*\\{', line, perl=TRUE)
    hit <- regmatches(line, m)[[1]]
    if (length(hit) == 2 && grepl("/", hit[[2]], fixed=TRUE)) { append_current(); current <- hit[[2]] }
    b <- regexec('"bytes"[[:space:]]*:[[:space:]]*([0-9]+)', line, perl=TRUE)
    bh <- regmatches(line, b)[[1]]
    if (length(bh) == 2 && !is.na(current)) { current_bytes <- as.numeric(bh[[2]]); append_current() }
    h <- regexec('"sha256"[[:space:]]*:[[:space:]]*"([0-9a-f]{64})"', line, perl=TRUE)
    hh <- regmatches(line, h)[[1]]
    if (length(hh) == 2 && !is.na(current)) { current_hash <- hh[[2]]; if (!is.na(current_bytes)) append_current() }
  }
  append_current()
  data.frame(rel=rel, bytes=bytes, sha256=hashes, stringsAsFactors=FALSE)
}

portable_manifest_match <- function(path, rel, expected) {
  if (!tolower(tools::file_ext(path)) %in% c("csv", "json", "md", "txt", "yml", "yaml", "svg")) return(FALSE)
  accepted_tmp <- tempfile(fileext=tools::file_ext(path)); error_tmp <- tempfile()
  on.exit(unlink(c(accepted_tmp, error_tmp)), add=TRUE)
  status <- system2("git", c("-C", root, "show", paste0("HEAD:", rel)), stdout=accepted_tmp, stderr=error_tmp)
  if (!is.null(status) && status != 0) return(FALSE)
  accepted_raw <- readBin(accepted_tmp, "raw", n=as.numeric(file.info(accepted_tmp)$size))
  working_raw <- readBin(path, "raw", n=as.numeric(file.info(path)$size))
  canonical <- function(x) charToRaw(gsub("\r\n?", "\n", rawToChar(x), perl=TRUE))
  if (!identical(canonical(accepted_raw), canonical(working_raw))) return(FALSE)
  lf_tmp <- tempfile(fileext=tools::file_ext(path)); crlf_tmp <- tempfile(fileext=tools::file_ext(path))
  on.exit(unlink(c(lf_tmp, crlf_tmp)), add=TRUE)
  writeBin(canonical(accepted_raw), lf_tmp)
  writeBin(charToRaw(gsub("\n", "\r\n", rawToChar(canonical(accepted_raw)), fixed=TRUE)), crlf_tmp)
  expected %in% c(sha256_file(lf_tmp), sha256_file(crlf_tmp))
}

portable_final_match <- function(path, expected, expected_bytes) {
  raw <- readBin(path, "raw", n=as.numeric(file.info(path)$size))
  ext <- tolower(tools::file_ext(path))
  if (!(ext %in% c("csv", "json", "md", "txt", "yml", "yaml", "svg"))) return(length(raw) == expected_bytes && sha256_file(path) == expected)
  canonical <- charToRaw(gsub("\r\n?", "\n", rawToChar(raw), perl=TRUE))
  crlf <- charToRaw(gsub("\n", "\r\n", rawToChar(canonical), fixed=TRUE))
  lf_tmp <- tempfile(fileext=paste0(".", ext)); crlf_tmp <- tempfile(fileext=paste0(".", ext))
  on.exit(unlink(c(lf_tmp, crlf_tmp)), add=TRUE)
  writeBin(canonical, lf_tmp); writeBin(crlf, crlf_tmp)
  hashes <- c(sha256_file(path), sha256_file(lf_tmp), sha256_file(crlf_tmp))
  expected %in% hashes && expected_bytes %in% c(length(raw), length(canonical), length(crlf))
}

verify_final <- function(i) {
  path <- file.path(reports, final_names[[i]])
  stopifnot(file.exists(path))
  txt <- paste(readLines(path, warn=FALSE), collapse="\n")
  stopifnot(grepl(paste0('"accepted_phase": "', expected_phase[[i]], '"'), txt, fixed=TRUE))
  stopifnot(grepl('"status": "ACCEPTED / FROZEN"', txt, fixed=TRUE))
  stopifnot(grepl(paste0('"source_commit": "', source_commit, '"'), txt, fixed=TRUE))
  stopifnot(grepl('"accepted_date": "2026-09-04"', txt, fixed=TRUE))
  for (name in names(expected_counts[[i]])) stopifnot(grepl(paste0('"', name, '": ', expected_counts[[i]][[name]]), txt, fixed=TRUE))
  x <- manifest_artifacts(path)
  stopifnot(nrow(x) == length(expected_artifacts[[i]]), setequal(x$rel, expected_artifacts[[i]]))
  for (j in seq_len(nrow(x))) {
    p <- file.path(root, x$rel[[j]])
    stopifnot(portable_final_match(p, x$sha256[[j]], x$bytes[[j]]))
  }
  x$rel
}

all_rel <- character(); entries <- 0L
for (i in seq_along(final_names)) {
  rel <- verify_final(i)
  all_rel <- c(all_rel, rel); entries <- entries + length(rel)
}
stopifnot(length(unique(all_rel)) == 26L, entries == 30L)

b_txt <- paste(readLines(file.path(reports, final_names[[2]]), warn=FALSE), collapse=" ")
a_path <- file.path(reports, final_names[[1]])
stopifnot(grepl(paste0('"sha256": "', sha256_file(a_path), '"'), b_txt, fixed=TRUE), grepl(paste0('"bytes": ', file.info(a_path)$size), b_txt, fixed=TRUE))

prior_total <- 0L
prior_names <- list.files(reports, pattern="^phase.*_freeze_manifest[.]json$", full.names=FALSE)
prior_names <- prior_names[!(prior_names %in% c(final_names, "phase10c_governance_futures_freeze_manifest.json"))]
for (name in prior_names) {
  x <- manifest_artifacts(file.path(reports, name))
  stopifnot(nrow(x) > 0)
  for (i in seq_len(nrow(x))) {
    p <- file.path(root, x$rel[[i]])
    stopifnot(file.exists(p))
    exact <- sha256_file(p) == x$sha256[[i]]
    portable <- if (exact) TRUE else portable_manifest_match(p, x$rel[[i]], x$sha256[[i]])
    stopifnot(isTRUE(exact || portable))
  }
  prior_total <- prior_total + nrow(x)
}
stopifnot(prior_total == 196L)

post <- paste(readLines(file.path(reports, "governance_post_correction_independent_review.md"), warn=FALSE), collapse=" ")
additional <- paste(readLines(file.path(reports, "governance_additional_post_correction_independent_review.md"), warn=FALSE), collapse=" ")
stopifnot(grepl("deleg_2e8d515e", post, fixed=TRUE), grepl('"passed": true', post, fixed=TRUE), grepl("deleg_eedc4116", additional, fixed=TRUE), grepl('"passed": true', additional, fixed=TRUE))
cat(sprintf("Phase 10 freeze R validation passed: %d final manifests, %d manifest entries, %d unique Phase 10 artifacts, %d prior Phase 1-9 protected artifacts; review records preserved\n", length(final_names), entries, length(unique(all_rel)), prior_total))
